✨Bento `amp-soundcloud`

AnuragVasanwala opened the pull request:
- [x] Add `amp-soundcloud`
- [x] Add types
- [x] Add type annotations
- [x] Add test case
- [x] Update Storybook
- [x] latestVersion precedes options
- [x] prettify new `amp-soundcloud`

Closes #34734
<!--
# Instructions:

1. Pick a meaningful title for your pull request.
  a. Prefix the title with an emoji. (Copy-paste from the list below.)
  b. If helpful, use a short prefix (e.g. `[Project XX] Implement feature YY`).
2. Enter a description that explains why the PR is necessary, and what it does.
  a. Mention the GitHub issue being addressed by this pull request.
  b. Use keywords to auto-close linked issues during merge. (e.g. `Fixes #11111`, or `Closes #22222`)
3. For substantial changes, first file an Intent-to-Implement (I2I) issue at go.amp.dev/i2i.

# References:

- AMP code contribution docs: go.amp.dev/contribute/code
- First time setup (required for CI checks): go.amp.dev/getting-started

# Emojis for categorizing pull requests (copy-paste emoji into description):

✨ New feature
🐛 Bug fix
🔥 P0 fix
✅ Tests
❄️ Flaky tests
🚀 Performance improvements
🖍 CSS / Styling
♿ Accessibility
🌐 Internationalization
📖 Documentation
🏗 Infrastructure / Tooling / Builds / CI
⏪ Revert
♻️ Refactor
🚮 Deletion
🧪 Experimental code
-->

AnuragVasanwala committed:
Initial Commit

AnuragVasanwala committed:
Property mappings and Cleanup

AnuragVasanwala committed:
Minor fixes, Formatting and Comments

AnuragVasanwala committed:
More readable variable name

AnuragVasanwala committed:
Test Cases for Bento

AnuragVasanwala committed:
Updated test-cases and Storybook

AnuragVasanwala committed:
Merge branch 'ampproject:main' into bento/amp-soundcloud

AnuragVasanwala committed:
Added base CSS File

AnuragVasanwala committed:
Merge branch 'bento/amp-soundcloud' of https://github.com/rtCamp/amphtml into bento/amp-soundcloud

ediamin commented on extensions/amp-soundcloud/1.0/storybook/Basic.amp.js:
```diff
@@ -0,0 +1,136 @@
+  return (
+    <amp-soundcloud
+      color={hex}
+      width="100%"
```
@AnuragVasanwala I'm getting error in console for this percentage width, `Invalid "width" value:  100% `.

pradeep910 replied:
I think percentage width should not be used. The element will need a fixed width & height. Right?

ediamin replied:
I'm trying to find out width percentage usage in existing any amp storybook js files but cannot getting any. According to this line https://github.com/rtCamp/amphtml/blob/main/src/layout.js#L399 we have to use `px` or `auto` for width attribute.

AnuragVasanwala replied:
Both of you are correct.

Here's, my analysis for the [`parseLength`](https://github.com/rtCamp/amphtml/blob/b2cd4622955e4ec8eef9643e231184b2adde0b5e/src/layout.js#L192).

#### Analysis
`height` parameter can have [`fluid`](https://github.com/rtCamp/amphtml/blob/b2cd4622955e4ec8eef9643e231184b2adde0b5e/src/layout.js#L408) as value or any value from below list:
`width` can have [`auto`](https://github.com/rtCamp/amphtml/blob/b2cd4622955e4ec8eef9643e231184b2adde0b5e/src/layout.js#L399) as value or any value from below list:
- [Number](https://github.com/rtCamp/amphtml/blob/b2cd4622955e4ec8eef9643e231184b2adde0b5e/src/layout.js#L193)
  - Example: `height=100`, `width=200`
- [String containing Relative and OtherLengthUnits](https://github.com/rtCamp/amphtml/blob/b2cd4622955e4ec8eef9643e231184b2adde0b5e/src/layout.js#L199) (px|em|rem|vh|vw|vmin|vmax|cm|mm|q|in|pc|pt)
  - Example: `height="100px"`, `width="100vw"`
- [String containing Number](https://github.com/rtCamp/amphtml/blob/b2cd4622955e4ec8eef9643e231184b2adde0b5e/src/layout.js#L202)
  - Example: `height="100"`, `width="200"`
- Values other than above will be rejected ([width](https://github.com/rtCamp/amphtml/blob/b2cd4622955e4ec8eef9643e231184b2adde0b5e/src/layout.js#L400) value check, [height](https://github.com/rtCamp/amphtml/blob/b2cd4622955e4ec8eef9643e231184b2adde0b5e/src/layout.js#L409) value check)
  - Example: `height="100%"`, `width="100%"`, `height="auto"`, `width="fluid"`

> If input is number, it will be converted to `px` by default for both [number](https://github.com/rtCamp/amphtml/blob/b2cd4622955e4ec8eef9643e231184b2adde0b5e/src/layout.js#L194) and [stringNumber](https://github.com/rtCamp/amphtml/blob/b2cd4622955e4ec8eef9643e231184b2adde0b5e/src/layout.js#L203).

Thank you @pradeep910 and @ediamin to provide suggestion 🙂
I am updating as per the above research.

AnuragVasanwala replied:
Any suggestion for this error:
<img width="1264" alt="image" src="https://user-images.githubusercontent.com/75766877/121341414-43f1af80-c93e-11eb-9eed-9f5716432858.png">

This error only comes if there are more than one `amp-soundcloud` in [MediaQuery - Test](https://github.com/rtCamp/amphtml/blob/773e788c7ae429aa5fcefe459f0ed10b13bec9eb/extensions/amp-soundcloud/1.0/storybook/Basic.amp.js#L78).

**Update : Possible Problem**
Problem is associated with `media` query. Whenever an `iframe` goes out of view (due to `media` query), some image in _Soundcloud_ `iframe` is trying to resize or maybe trying to reload. And because `iframe` is no longer in view, its actual/rendered width become `ZERO`.

I think, this is not a critical issue.

ediamin reviewed (commented):
@AnuragVasanwala  I'm getting these following errors in my console in Storybook for amp window:
![Screenshot 2021-06-09 at 1 43 58 PM](https://user-images.githubusercontent.com/1541774/121313984-c1aac080-c928-11eb-8b9a-c951b618e138.png)

ediamin commented on extensions/amp-soundcloud/1.0/base-element.js:
```diff
@@ -0,0 +1,64 @@
+  'color': {attr: 'data-color'},
+  'height': {attr: 'height'},
+  'heights': {attr: 'heights'},
+  'layout': {attr: 'layout'},
```
We can skip this `layout` prop here and remove it from this file. This attribute is handled by the layout.js here https://github.com/rtCamp/amphtml/blob/main/src/layout.js#L383.

AnuragVasanwala committed:
Bug Fix: Layout System and Unit

AnuragVasanwala committed:
Parent element changed from <div> to <>

AnuragVasanwala committed:
Type correction for 'visual' parameter

AnuragVasanwala committed:
Merge branch 'ampproject:main' into bento/amp-soundcloud

AnuragVasanwala committed:
Bug fix for color parameter in storybook

AnuragVasanwala committed:
Merge branch 'bento/amp-soundcloud' of https://github.com/rtCamp/amphtml into bento/amp-soundcloud

AnuragVasanwala committed:
Update extensions/amp-soundcloud/1.0/base-element.js

Bento components are required to be R1. We have an equivalent static getPreconnects(element) in R1 designed to take care of this.

Co-authored-by: Justin Ridgewell <justin@ridgewell.name>

AnuragVasanwala committed:
♻️ Add 'wrapper' designation to bento

#34838

Co-authored-by: Caroline Liu <10456171+caroqliu@users.noreply.github.com>

AnuragVasanwala committed:
🚮 Delete init/constructor of `amp-soundcloud`

No need to override `init` if no changes were made to it except calling parent class's `init`.

Co-authored-by: Caroline Liu <10456171+caroqliu@users.noreply.github.com>

AnuragVasanwala committed:
♻️ Use `passthroughNonEmpty` which only give `<slot>` for non empty `children`

Since `passthrough` gives a `<slot>` reference to the children prop in all cases, let's make this passthroughNonEmpty -- which will only give a `<slot>` if there are children in the given markup.

Co-authored-by: Caroline Liu <10456171+caroqliu@users.noreply.github.com>

AnuragVasanwala committed:
♻️ Warning for required props when not present

AnuragVasanwala committed:
♻️ Warning for required props when not present

AnuragVasanwala committed:
♻️ Warning for required pros when not present

AnuragVasanwala committed:
Merge branch 'bento/amp-soundcloud' of https://github.com/rtCamp/amphtml into bento/amp-soundcloud

AnuragVasanwala committed:
♻️ `trackId` and `playlistId` order chek

Co-authored-by: Caroline Liu <10456171+caroqliu@users.noreply.github.com>

AnuragVasanwala committed:
♻️ `trackId` and `playlistId` order check

Co-authored-by: Caroline Liu <10456171+caroqliu@users.noreply.github.com>

AnuragVasanwala committed:
Merge branch 'bento/amp-soundcloud' of https://github.com/rtCamp/amphtml into bento/amp-soundcloud

AnuragVasanwala committed:
♻️ Removed explicitly specify boolean `true` for conditional block

No need to explicitly specify boolean `true` for conditional block.

Co-authored-by: Caroline Liu <10456171+caroqliu@users.noreply.github.com>

AnuragVasanwala committed:
♻️ Added default value for `visual` prop in Preact component

Co-authored-by: Caroline Liu <10456171+caroqliu@users.noreply.github.com>

AnuragVasanwala committed:
🚮 Removed props handled by AMP's layout system

`height`, `heights`, `media`, `sizes`, and `width` are not needed to pass through as these attributes are specific to AMP's layout system and handled by the runtime.

Co-authored-by: Caroline Liu <10456171+caroqliu@users.noreply.github.com>

AnuragVasanwala committed:
🚮 Removed unnecessary typedef

`height`, `heights`, `media`, `sizes`, and `width` are not needed to pass through as these attributes are specific to AMP's layout system and handled by the runtime. Thus they do not need to be specify in typedef.

Co-authored-by: Caroline Liu <10456171+caroqliu@users.noreply.github.com>

AnuragVasanwala committed:
🚮 Removed interface as no API exposed

AnuragVasanwala committed:
♻️ Moved `converter.js` into storybook

AnuragVasanwala committed:
🚮 Removed `layout` from Preact example, ♻️ Added `height` into style

AnuragVasanwala committed:
🚮 Removed empty CSS and `hasCss` from `bundles.config.extensions.json`

AnuragVasanwala committed:
🐛 `htmlFor` import fix for test case, ♻️ added `data-trackid` attribute

AnuragVasanwala committed:
♻️ Replaced `<iframe>` with `<IframeEmbed>`

AnuragVasanwala committed:
♻️ Updated specification for validation

AnuragVasanwala committed:
🏗 Added `npm` to `bundles.config.extensions.json`

Co-authored-by: Caroline Liu <10456171+caroqliu@users.noreply.github.com>

AnuragVasanwala committed:
♻️ New line at EOF in `validator-amp-soundcloud.out`

AnuragVasanwala committed:
Added 'loadable' to indicate a resource intensive component

AnuragVasanwala committed:
🏗 Fixes reported by 'amp lint' for CI

AnuragVasanwala committed:
Merge branch 'ampproject:main' into bento/amp-soundcloud

AnuragVasanwala committed:
🏗 Fixes for aliased imports for CI
